### PR TITLE
(t) data-collector (Dashboard backend) fails to start #2368

### LINKED ISSUE #2368: (t) data-collector (Dashboard backend) fails to start
(t) = testing branch.Our meta issue of #2362 has an entry for mostly failing dashboard. This is mostly likely down to the known failure of our data-collector system to start. This also affects numerous Web-UI 'live' header info. That to @FroggyFlox for doing some initial investigation into this issue that has arrisen post our recent Django update in the testing branch and soon to be testing channel releases.

From: less /opt/rockstor-dev/var/log/supervisord_data-collector_stderr.log

```
Traceback (most recent call last):
  File "/opt/rockstor-dev/bin/data-collector", line 31, in <module>
    import smart_manager.data_collector
  File "/opt/rockstor-dev/src/rockstor/smart_manager/data_collector.py", line 44, in <module>
    from system.pinmanager import (
  File "/opt/rockstor-dev/eggs/gevent-1.1.2-py2.7-linux-x86_64.egg/gevent/builtins.py", line 93, in __import__
    result = _import(*args, **kwargs)
  File "/opt/rockstor-dev/src/rockstor/system/pinmanager.py", line 23, in <module>
    from storageadmin.models import Pincard, EmailClient, User
  File "/opt/rockstor-dev/eggs/gevent-1.1.2-py2.7-linux-x86_64.egg/gevent/builtins.py", line 93, in __import__
    result = _import(*args, **kwargs)
  File "/opt/rockstor-dev/src/rockstor/storageadmin/models/__init__.py", line 19, in <module>
    from pool import Pool  # noqa E501
  File "/opt/rockstor-dev/eggs/gevent-1.1.2-py2.7-linux-x86_64.egg/gevent/builtins.py", line 93, in __import__
    result = _import(*args, **kwargs)
  File "/opt/rockstor-dev/src/rockstor/storageadmin/models/pool.py", line 34, in <module>
    class Pool(models.Model):
  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/db/models/base.py", line 110, in __new__
    app_config = apps.get_containing_app_config(module)
  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/apps/registry.py", line 247, in get_containing_app_config
    self.check_apps_ready()
  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/apps/registry.py", line 125, in check_apps_ready
    raise AppRegistryNotReady("Apps aren't loaded yet.")
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
```

### PR DESCRIPTION
**Pull request by phillxnet**
Add in required as of Django 1.11 django.setup() to our data_collector. It may-well be that we are still doing things wrong here but adding for now to re-establish some more basic functionality.

N.B. django.setup() can only be run once. See included reference.

Fixes #2368 

## Testing
Post patch we successfully start our data-collector via supervisord:
```
2022-04-11 18:44:54,422 CRIT Supervisor running as root (no user in config file)
2022-04-11 18:44:54,429 INFO RPC interface 'supervisor' initialized
2022-04-11 18:44:54,429 CRIT Server 'unix_http_server' running without any HTTP authentication checking
2022-04-11 18:44:54,430 INFO supervisord started with pid 23283
2022-04-11 18:44:55,433 INFO spawned: 'nginx' with pid 23287
2022-04-11 18:44:55,434 INFO spawned: 'gunicorn' with pid 23288
2022-04-11 18:44:55,437 INFO spawned: 'data-collector' with pid 23289
2022-04-11 18:44:55,440 INFO spawned: 'ztask-daemon' with pid 23290
2022-04-11 18:44:58,385 INFO success: data-collector entered RUNNING state, process has stayed up for > than 2 seconds (startsecs)
2022-04-11 18:45:01,389 INFO success: nginx entered RUNNING state, process has stayed up for > than 5 seconds (startsecs)
2022-04-11 18:45:01,390 INFO success: gunicorn entered RUNNING state, process has stayed up for > than 5 seconds (startsecs)
2022-04-11 18:45:06,397 INFO success: ztask-daemon entered RUNNING state, process has stayed up for > than 10 seconds (startsecs)
```
Post pr we have no additional failures (for current testing branch) on our tests:
```
Ran 208 tests in 19.124s
FAILED (failures=4, errors=10)
```
And a working dashboard and 'live' Web-UI elements.

### CODE REVIEW

**Comment by phillxnet** [comment]
There is some reservation as to the 'fix' provided here. It may well be we are still working around some of our existing Django customisation. I.e. Note our conf/django-hack.

See also the reference for this 'fix' (also recorded in code comments within this pull request):
https://docs.djangoproject.com/en/1.11/topics/settings/#calling-django-setup-is-required-for-standalone-django-usage

However we have a long way to go yet re our overarching #1877 which will in turn allow for our next major Django update.
Further clean-up can be addressed along the way.

**Comment by phillxnet** [comment]
Thanks to @FroggyFlox for pointing out the following:
*********************
Apparently we already have two prior instances of django-setup():
https://github.com/rockstor/rockstor-core/blob/eeb0084a2ea6f6bdb1cdbe35428aa5021ddadc72/src/rockstor/scripts/__init__.py
For our stand alone scripts.
And
https://github.com/rockstor/rockstor-core/blob/eeb0084a2ea6f6bdb1cdbe35428aa5021ddadc72/src/rockstor/smart_manager/replication/__init__.py
presumably for our replication tasks.

The former commit indicates a similar requirement 'back then' re a much older Django update:
https://github.com/rockstor/rockstor-core/commit/2038c04bbbbb9d7b21159935b7588be06292d217

> djagon app registry setup refactor. Fixes #1560
> We moved to django 1.8 recently from 1.6. As of 1.7, however, we have to
> explicitly populate app registry. Since most scripts, if not all use the ORM,
> do the django.setup in scripts.__init__.py

*********************

**Comment by phillxnet** [comment]
Noting a reference from:
https://github.com/rockstor/rockstor-core/pull/1557#issuecomment-264669063
Relating to the Django 1.7 redesign of the “app registry”

We also have the following which looks very much like what we are addressing here:
https://github.com/rockstor/rockstor-core/issues/1560

Our prior pincardmanager user of django.setup() for model access was removed via:
"Fixing Issue#1567 datacollector orm to api migration #1569"
https://github.com/rockstor/rockstor-core/pull/1569

Our prior pincardmanager user of django.setup() was removed via:
"Fixing Issue#1567 datacollector orm to api migration #1569"
https://github.com/rockstor/rockstor-core/pull/1569

So to summarise, post the addition proposed in this patch, we have the following current instances of django-setup()
```
grep --exclude-dir=venv -R "django.setup()" *

src/rockstor/scripts/__init__.py:django.setup()
src/rockstor/smart_manager/replication/__init__.py:django.setup()
src/rockstor/smart_manager/data_collector.py:django.setup()
```

**Comment by phillxnet** [comment]
Testing one of our scripts that does model access write:
Installed Netdata Rock-on and tested as working OK.
Turned off Netdata Rock-on.
Viewed Dashboard live.
Wiped associated db entries via 'emergency maintenance' script via command line as follows:
```
rleap15-3:~ # /opt/rockstor-dev/bin/delete-rockon "Netdata (official)"
/opt/rockstor-dev/eggs/psycopg2-2.7.4-py2.7-linux-x86_64.egg/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
Rock-On(Netdata (official)) metadata in the db is deleted
```
with the corresponding entries in rockstor.log:
```
[18/Apr/2022 18:43:45] ERROR [storageadmin.views.network:213] NetworkConnection matching query does not exist.
Traceback (most recent call last):
  File "/opt/rockstor-dev/src/rockstor/storageadmin/views/network.py", line 207, in update_connection
    name=dconfig["connection"]
  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/db/models/query.py", line 380, in get
    self.model._meta.object_name
DoesNotExist: NetworkConnection matching query does not exist.
[18/Apr/2022 18:43:59] DEBUG [system.osi:206] Running command: /usr/bin/docker stop netdata_official
[18/Apr/2022 18:43:59] DEBUG [system.osi:206] Running command: /usr/bin/docker rm netdata_official
[18/Apr/2022 18:43:59] DEBUG [system.osi:206] Running command: /usr/bin/docker rmi netdata/netdata:latest
...
[18/Apr/2022 18:44:45] DEBUG [storageadmin.views.share_helpers:113] Updating pre-existing same pool db share entry.
[18/Apr/2022 18:44:45] DEBUG [system.osi:206] Running command: /usr/sbin/btrfs subvolume list /mnt2/ROOT
[18/Apr/2022 18:44:46] DEBUG [storageadmin.views.share_helpers:249] Removing, missing on disk, snapshot db entry (34fa1df54da2ad50f68f100c09d224c19feff4d9e04675e398a94916666ce5a3) from share (rock-ons-root).
[18/Apr/2022 18:44:46] DEBUG [storageadmin.views.share_helpers:249] Removing, missing on disk, snapshot db entry (410810dbd8113a8fbab32e09c09d6fa8f509bdb8c8bac20bdc66bdf3a2285647) from share (rock-ons-root).
[18/Apr/2022 18:44:46] DEBUG [storageadmin.views.share_helpers:249] Removing, missing on disk, snapshot db entry (98867a95898e5c483310833b1a92616a887e3b91b161e3aa31afe67404641d20) from share (rock-ons-root).
[18/Apr/2022 18:44:46] DEBUG [storageadmin.views.share_helpers:249] Removing, missing on disk, snapshot db entry (3a309b5e5932b6f2df594a696d1e5160741b452332c2e5867c536fd5d8ba3d6c) from share (rock-ons-root).
[18/Apr/2022 18:44:46] DEBUG [storageadmin.views.share_helpers:249] Removing, missing on disk, snapshot db entry (f7768d4045f124b72086379e21032c857ebed84569099acb3083235891585928) from share (rock-ons-root).
[18/Apr/2022 18:44:46] DEBUG [storageadmin.views.share_helpers:249] Removing, missing on disk, snapshot db entry (e98eb917cd5c86953ec7ecf3423c0da7dd731d55ca12a953948b8360324318ec) from share (rock-ons-root).
[18/Apr/2022 18:44:46] DEBUG [storageadmin.views.share_helpers:249] Removing, missing on disk, snapshot db entry (414b69d7b316054f25f3e963da6109c8599b2d0dfe5e524461b088040d59ee46) from share (rock-ons-root).
[18/Apr/2022 18:44:46] DEBUG [storageadmin.views.share_helpers:249] Removing, missing on disk, snapshot db entry (34fa1df54da2ad50f68f100c09d224c19feff4d9e04675e398a94916666ce5a3-init) from share (rock-ons-root).
```
So at least this stand-along script looks to be functioning OK with db access and write (to remove the Rock-on entries via ro.delete()).
A subsequent view of the Rock-ons page showed no Rock-ons installed.

**Comment by phillxnet** [comment]
We can't as yet test replication's own invocation of django.setup() via inherriting scripts __init__  due to the System -> Services page failure blocking a test setup of replication. The following issue has been created to address this blocker:
(t) System Services page fails #2370

**Comment by phillxnet** [comment]
@FroggyFlox further to the testing here of at least our docker-delete stand alone script executing as expected, with it's own implicit django-setup() via the associated scrip dir __init__.py I'm going to merge this pull request as-is to aid further collaboration of our outstanding recent Django updates issues. I'll update the meta issue for this area of work now directly.
